index/scorch/merge.go

```diff
@@ -129,8 +147,116 @@ func (s *Scorch) parseMergePlannerOptions() (*mergeplan.MergePlanOptions,
 	return &mergePlannerOptions, nil
 }
 
+type mergerCtrl struct {
+	creator string
+	options *mergeplan.MergePlanOptions
+	doneCh  chan struct{}
+}
+
+// MergeRequest represents various control
+// parameters for the RequestMerge API.
+type MergeRequest struct {
+	// MergeOptions specify the merge policy applied during
+	// the forced merge cycles. This doesn't override the
+	// index's original merge policy.
+	MergeOptions *mergeplan.MergePlanOptions
+
+	// OverrideNumSnapshotsToKeep specify whether to retain
+	// a number of older snapshots dictated by numSnapshotsToKeep
+	// during a forced merge cycle. Enabling this reduces the
+	// disk space requirements during a forced merge operation.
+	OverrideNumSnapshotsToKeep bool
+
+	// DoneCh indicates the completion of the asynchronous
+	// forced merge operation.
+	DoneCh chan struct{}
+}
+
+func (s *Scorch) RequestMerge(mr *MergeRequest) error {
+	// check whether force merge is already under processing
+	if atomic.LoadUint64(&s.stats.TotFileMergeForceOpsStarted) >
```

Comment:
I'm wondering why you still do this (ineffective in my opinion) check at the top, and then do the safe check (using lock) inside the goroutine you've spawned.  By doing it this way, you have to introduce the error channel and more complex error return handling code.  Couldn't you have placed the safe check (using lock) at the top of this function, and then you no longer need to check again inside the goroutine, as you've already ensured that any concurrent invocations have been stopped early.  Then the need for the error channel goes away too.

Reply:
hm, my worry there was - since the go routine spin ups and scheduling is an async process there could be many threads passing the first check and may end up seeing green signals for spinning up go routines until the first routine really goes gets to the execution to bump those stats..?